File: frontend/src/api/comments.ts
import { api } from "./client";

export interface CommentAuthor {
  id: string;
  display_name: string;
}

export interface Comment {
  id: string;
  documantions_id: string;
  author_id: string;
  author_name?: string;
  author?: CommentAuthor;
  body: string;
  created_at: string;
  updated_at?: string | null;
}

export interface CommentRaw {
  id: string;
  documantions_id?: string;
  author_id: string;
  author_display_name?: string | null;
  author_name?: string;
  author?: CommentAuthor;
  body: string;
  created_at: string;
  edited_at?: string | null;
  updated_at?: string | null;
}

export interface CommentListEnvelope {
  documantions_id: string;
  comments: CommentRaw[];
  total: number;
  skip: number;
  limit: number;
}

export interface CommentListParams {
  limit?: number;
  skip?: number;
  offset?: number;
  sort?: "newest" | "oldest" | "created_at" | "-created_at";
}

function backendSort(sort: CommentListParams["sort"]): "newest" | "oldest" | undefined {
  if (sort === "created_at") return "oldest";
  if (sort === "-created_at") return "newest";
  return sort;
}

function toQuery(params: CommentListParams = {}): string {
  const search = new URLSearchParams();
  if (params.limit !== undefined) search.set("limit", String(params.limit));
  const skip = params.skip ?? params.offset;
  if (skip !== undefined) search.set("skip", String(skip));
  const sort = backendSort(params.sort);
  if (sort) search.set("sort", sort);
  const query = search.toString();
  return query ? `?${query}` : "";
}

function mapComment(raw: CommentRaw, docId: string): Comment {
  return {
    id: raw.id,
    documantions_id: raw.documantions_id ?? docId,
    author_id: raw.author_id,
    author_name: raw.author_display_name ?? raw.author_name,
    author: raw.author,
    body: raw.body,
    created_at: raw.created_at,
    updated_at: raw.edited_at ?? raw.updated_at ?? null,
  };
}

export async function listComments(docId: string, params?: CommentListParams): Promise<Comment[]> {
  const envelope = await api.get<CommentListEnvelope>(`/documents/${docId}/comments${toQuery(params)}`);
  return envelope.comments.map((comment) => mapComment(comment, envelope.documantions_id));
}

export async function createComment(docId: string, body: string): Promise<Comment> {
  const comment = await api.post<CommentRaw>(`/documents/${docId}/comments`, { body });
  return mapComment(comment, docId);
}

export async function updateComment(docId: string, commentId: string, body: string): Promise<Comment> {
  const comment = await api.patch<CommentRaw>(`/documents/${docId}/comments/${commentId}`, { body });
  return mapComment(comment, docId);
}

export function deleteComment(docId: string, commentId: string): Promise<void> {
  return api.delete<void>(`/documents/${docId}/comments/${commentId}`);
}

File: src/services/api/routers/comments.py
from __future__ import annotations

from typing import Annotated, Any
from uuid import UUID

from fastapi import APIRouter, Depends, HTTPException, Request

from services.api._helpers import _fmt_dt
from services.api.main import current_user
from services.auth.models import TokenPayload
from services.auth.repository import AuthRepository
from services.comments.models import CommentCreateRequest, CommentUpdateRequest
from services.comments.repository import CommentRepository
from services.permissions.enforcer import assert_doc_access
from shared.db import to_uuid

router = APIRouter(tags=["comments"])


@router.get("/documents/{document_id}/comments")
def list_comments(
    document_id: UUID,
    request: Request,
    user: Annotated[TokenPayload, Depends(current_user)],
    skip: int = 0,
    limit: int = 50,
    sort: str = "newest",
) -> dict[str, Any]:
    with request.app.state.engine.begin() as connection:
        auth_repo = AuthRepository(connection)
        assert_doc_access(document_id, user, auth_repo)

        repo = CommentRepository(connection)
        comments = repo.list_comments(document_id, skip=skip, limit=limit, sort=sort)
        total = repo.count_comments(document_id)
        return {
            "document_id": str(document_id),
            "comments": [
                {
                    "id": str(to_uuid(c["id"])),
                    "author_id": str(to_uuid(c["author_id"])),
                    "author_display_name": c["author_display_name"],
                    "body": c["body"],
                    "created_at": _fmt_dt(c["created_at"]),
                    "edited_at": _fmt_dt(c["edited_at"]),
                    "edited_by_id": (
                        str(to_uuid(c["edited_by_id"])) if c["edited_by_id"] else None
                    ),
                    "deleted_at": _fmt_dt(c["deleted_at"]),
                }
                for c in comments
            ],
            "total": total,
            "skip": skip,
            "limit": limit,
        }


@router.post("/documents/{document_id}/comments", status_code=201)
def create_comment(
    document_id: UUID,
    body: CommentCreateRequest,
    request: Request,
    user: Annotated[TokenPayload, Depends(current_user)],
) -> dict[str, Any]:
    with request.app.state.engine.begin() as connection:
        auth_repo = AuthRepository(connection)
        assert_doc_access(document_id, user, auth_repo)

        repo = CommentRepository(connection)
        comment = repo.create(document_id, user.sub, body.body)
        request.app.state.metrics.comments_total.labels("create", "success").inc()
        return {
            "id": str(to_uuid(comment["id"])),
            "document_id": str(to_uuid(comment["document_id"])),
            "author_id": str(to_uuid(comment["author_id"])),
            "body": comment["body"],
            "created_at": _fmt_dt(comment["created_at"]),
        }


@router.patch("/documents/{document_id}/comments/{comment_id}")
def update_comment(
    document_id: UUID,
    comment_id: UUID,
    body: CommentUpdateRequest,
    request: Request,
    user: Annotated[TokenPayload, Depends(current_user)],
) -> dict[str, Any]:
    with request.app.state.engine.begin() as connection:
        auth_repo = AuthRepository(connection)
        assert_doc_access(document_id, user, auth_repo)

        repo = CommentRepository(connection)
        comment = repo.get_by_id(comment_id)
        if comment is None or comment["deleted_at"] is not None:
            raise HTTPException(status_code=404, detail="Comment not found")
        if not repo.can_edit(comment_id, user.sub, user.is_admin):
            raise HTTPException(status_code=403, detail="Cannot edit this comment")

        repo.update(comment_id, body.body, edited_by_id=user.sub)
        request.app.state.metrics.comments_total.labels("update", "success").inc()
        updated = repo.get_by_id(comment_id)
        if updated is None:
            raise HTTPException(status_code=404, detail="Comment not found")
        return {
            "id": str(to_uuid(updated["id"])),
            "author_id": str(to_uuid(updated["author_id"])),
            "author_display_name": updated["author_display_name"],
            "body": updated["body"],
            "created_at": _fmt_dt(updated["created_at"]),
            "edited_at": _fmt_dt(updated["edited_at"]),
            "edited_by_id": (
                str(to_uuid(updated["edited_by_id"])) if updated["edited_by_id"] else None
            ),
        }


@router.delete("/documents/{document_id}/comments/{comment_id}", status_code=204)
def delete_comment(
    document_id: UUID,
    comment_id: UUID,
    request: Request,
    user: Annotated[TokenPayload, Depends(current_user)],
) -> None:
    with request.app.state.engine.begin() as connection:
        auth_repo = AuthRepository(connection)
        assert_doc_access(document_id, user, auth_repo)

        repo = CommentRepository(connection)
        comment = repo.get_by_id(comment_id)
        if comment is None or comment["deleted_at"] is not None:
            raise HTTPException(status_code=404, detail="Comment not found")
        if not repo.can_delete(comment_id, user.sub, user.is_admin):
            raise HTTPException(status_code=403, detail="Cannot delete this comment")

        repo.soft_delete(comment_id, deleted_by_id=user.sub)
        request.app.state.metrics.comments_total.labels("delete", "success").inc()
